### ISSUE TITLE
Recursive type inferred correctly but invalid assignment does not throw error

### ISSUE THREAD

**Issue by halfmatthalfcat** [microsoft/TypeScript#47119]
# Bug Report

### 🔎 Search Terms

recursive types, variadic tuple types, conditional types, recursive conditional types, generic recursive conditional types

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about the search terms above yet yielded no obvious answers.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.2#code/PQKhCgAIUgVALAppADgewJYDsAuk0BmkASogMYCuATgM4YBuiAsgIYqQY2Q5qQBGyFnwA2yHpDJVELHIKwATSDRxU0WAObCAnty0o5+PgCtyeQoZNkcXcyyp8MKu1qgx5iFDniR5VBoixIegwWSFCcChRRXX1IAAoAaQBKbngZH0QCbEQuL2QAa0QdHD0c-CI8pQo+GkQARwoAs2NTLhYFMMgsrBZhIN7G+OIk1y60KlTkbCxEKgBbNGULUwA6UdGAMXHIRAAPFjmoxAAudehIAElIAHd2vHblq3LIAG9IAG1d46UVbHUAXW+b0+32Ufg0gNeHy+kCwFDmAiokL4aDQogeAF8sWcYFdrmgKMJFHlAu5uoIYshSJRaP5WCgADzvMF-AA0P3B6nZcIRs3+7JRaOkWAAfDi4EgODQaIMkFIpal0pUyGhDhhRBMVVQpFZtBwsARZrlJWCKFZqMgtIg8NckFhRixpRh1DNFA8WObev1hIMvOk+FIWPljVMDbMAmRENzeLNVLQwvKvKprlgVpBRgBlRDIKgARjCHSoACYxhM9gcjjQ1tBgOBwMBgHBSpAKDh1dxeOprZAkCxic3zBEjuASrEABLSeQM2A7XayBRcACC2pYWgZFCw+SwaBTIpFkAAvHBZ-P5Fx3vrDRMx+yVnfl1RV+vN9vd5B-pAAPyQMeQb4zRgqAAbnrRtYGbVt23ELs8BwFgoIHCpIlEEdm1geDhGnE8AjPSAHyfDctx3UV9yPGc9lPc8WxfYjbzvbArzgDD3y-Jj23-RBAJAhsm1iSC+nEKRANqTpRzEZDEFQ2JSGExAsIonClxXNdCNfUV2WIbCFzw5TnyIt8j3ef5SKgOB3gAclEDQvHMj8FO0gAGUzIFY4hnL-Y97Nw94GNmH9JzolZfImdD1X+dyXO-GSjTk0LhHZd4Jz7QLiGMiKPIA2YQNA3jkH40tUFUfQqBKSBCmKUoaCk5AEiKGcjxZDRIAAH1heFERapQtARNFsrEkhyGoOhGHpBlTJ48ZnWwL1yspLhTISLTcPwtdaq0WARVZcbGykGVhDbJr6AGMREP1GZ5kWZpLBwbbUk4BUpCiD1EEUa5HG8FUsFNKwXsee4uGuHNBtoMR4FUCh1G8SpvoiKRTOILaXJ4oSYueRaaF4a41tgZAyAeLUdRwPU+AodViUlNAWieGGLVusS2nlDdan+m4gZpESkwJSHWcTeB7qHaJ7s4poKF6PVzAcyArPULwFvzLylMfVa6tIgbZIZBIxRMlyElzCzpZsuy50UyAnJcpHG1tAJWfMxgqjISNpQIQk9R1C1anJ8HIfZKRYcCFVDlEWRKXcqL3J4mDuGY8wpDmeCsD+OaO0gN2QfwQD3O+XWlqo1TaMgO8grDCZdbi99w8bR1ahKyYo-be7zAJ0wyqKJO4m3PAZXlW0dADlB1T8vGsHMvBgp7cMBVbHwMAIK8milGVEBGc3WNL5iFZ0pWNZV9KeKr2ZYMlXtFAbogm6eWb+vbtBO4tG54F71V+41CR2hHy8-LlQQAcQYRhGXlerEkpTl1vuDe2N0orx4iTMmKdgbDWQJTa6BZFCfQOoMN6XhIBr3VJA82PEAACHdkAXFtoIYQGNuB+F+pg7w1JxhTmARrXMm0SD7lobXGgBxkCZENFYPBkUBrszpGwDW7JgQAAV9T+T7MwkUkJNIYnZDg4QYpAGAI4hndRLlNGzEzrCYWwE6ygAgOcAAwk-AempxiEz1MFGgpxzjAhhI1AEQJoT-nanydxIIOR-GRKidEgQsQYlGPAHchj2SVCQc3eQaAygdwkGoAg4w5jJxlGQbwoQ6AuhkHfdoihtw7G1NsfmYNiLVhALWT6SwqDfGpENERjJmS-A0NyLxVB2SuP5PwQJwpVYvFMrmIsABmdxAAWAArAANncQAdgABwAE5vgEF6CJDEkBQkYiAkAA)

### 💻 Code

```ts
// Type util to get head type of tuple
type Head<T extends Array<unknown>> = T extends [ infer H, ...Array<unknown> ] ? H : never;
// Type util to get tail type of tuple
type Tail<T extends Array<unknown>> = T extends [ unknown, ...infer Tail ] ? Tail : never;
// Type util to reverse a type tuple
type Reverse<T extends Array<unknown>, R extends Array<unknown> = []> =
  T['length'] extends 0
    ? R
    : T extends [infer Head, ...infer Tail]
      ? Reverse<Tail, [Head, ...R]>
      : never;

// Type util for property key types
type KeyT = string | number | symbol;

type RecursiveMap<
  // original key types 
  K extends Array<KeyT>,
  // resulting value type of innermost object
  // this is replaced with constructed object as we recurse through the structure
  R,
  // reverse of K so we can correctly build the object structure
  // types are unset as we recurse through, where this tuple is eventually of 0 length
  K1 extends Array<KeyT> = Reverse<K>
> =
  K1['length'] extends 0
    // when we've successfully recursed through, return complete type
    ? R
    // get tail of remaining types to recurse over
    : K1 extends [ unknown, ...infer K1Tail ]
    // assert the tail is of correct key types (not sure why compiler can't infer here, but different issue)
    ? K1Tail extends Array<KeyT>
      // assert the head is of correct key type (not sure why compiler can't infer here as well)
      ? Head<K1> extends KeyT
          // build recursive object and continue with K1Tail
          // @note I've also tried with Record<Head<K1>, R> with the same effect
          ? RecursiveMap<K, { [P in Head<K1>]: R }, K1Tail>
          : never
        : never
    : never;

/**
 * Compiler correctly infers:
 * { [x: string]: { [x: number]: { [x: string]: boolean }}}
 * however, the object does not conform to such a signature and no error is thrown.
 */
const r: RecursiveMap<[string, number, string], boolean> = {
  123: { 456: { 789: false } }
};
```

### 🙁 Actual behavior

When assigning `{ 123: { 456: { 789: false } } }` to `r: RecursiveMap<[string, number, string], boolean>` no error is thrown that the shape of the object does not conform to the given type.

### 🙂 Expected behavior

Since the object's keys are not correct (`[number, number, number]` instead of `[string, number, string]`) the compiler should error on the incorrect assignment.

**Comment by MartinJohns** [comment]
>  * Compiler correctly infers:  
> * { [x: string]: { [x: number]: { [x: string]: boolean }}}

`number`s are always valid properties for a `string` record. They're coerced to `string`. See #30883 and this example:

    type Data = { [key: string]: boolean }
    const t: Data = { 12: true }

**Comment by halfmatthalfcat** [comment]
> > * Compiler correctly infers:
> > * { [x: string]: { [x: number]: { [x: string]: boolean }}}
> 
> `number`s are always valid properties for a `string` record. They're coerced to `string`. See #30883 and this example:
> 
> ```
> type Data = { [key: string]: boolean }
> const t: Data = { 12: true }
> ```

Ah I see. Thank you for the context!
